### ISSUE TITLE
Registering an SE key with an ID from the vendor range fails.

### ISSUE THREAD

**Issue by stevew817** [ARMmbed/mbedtls#3285]
### Description
- Type: Bug
-----------------------------------------------------------------

## Bug
When registering a key that already resides in an SE (accessed through an SE driver), one would expect that the PSA 'Vendor' key ID range would be allowed to identify such keys that are factory-provisioned (and thus cannot be created).
When calling `mbedtls_psa_register_se_key()`, its implementation calls `psa_validate_key_attributes()`, which in turn calls `psa_validate_persistent_key_parameters()` while always assuming that it needs to verify the parameters for a to-be-created key. This means registering an SE key with an ID in the vendor range will always fail the key attribute validation routine.

**Comment by gilles-peskine-arm** [comment]
Actually that's not just about `mbedtls_psa_register_se_key`. There's more than one entity that can count as the “vendor”: the manufacturer of the main chip, the manufacturer of a secure element, the OEM, a reseller that provisions initial keys, the provider of a trusted environment such as a TrustZone secure world, etc. Some of these may use the PSA API to create their keys, and yet their keys could count as “vendor”. The converse is possible as well: a system where the application use the PSA API but the key provisioning mechanism is proprietary, so the application calls `mbedtls_psa_register_se_key` to register its own key.

Or maybe this is the wrong direction, and “vendor” should be restricted to the entity that provides the core PSA crypto environment.

Or maybe the answer is a mix of both, and the vendor range should be subdivided between these entities. Given that the number and nature of such entity is highly variable, I think that would be an Mbed TLS design decision, not something that PSA would standardize. But this warrants further reflection and discussion.

**Comment by stevew817** [comment]
@gilles-peskine-arm we're getting to a point where we'd need to make a choice here. The PSA API defines a key ID range allocated to `vendor`, without specifying what a vendor is. Additionally, mbed TLS has now gone down a road where it is allowed to have multiple providers of an opaque driver on the same system.

I think the logical choice would be to implementation-define the meaning of the vendor ID range, and split it up between the providers of opaque drivers. I.e. an opaque driver is currently referenced by a 24-bit location value. The vendor key ID range is 31 bit. If we carve up the vendor key ID range using the location value, we can get 7 bits of keyspace (128 keys) per registered opaque driver.

Questions:
* Does that sound like the way to go to you?
* Would it make sense to reduce the location value breadth and allow for more keyspace per driver, but less registered drivers? The 23 bits of usable locations means we can have 8 million link-time compiled in drivers... Isn't that a bit much?
* Or do you see another way to go about this?

**Comment by stevew817** [comment]
I might have miscounted the bits in my previous comment. To summarize my proposal:

```
psa_key_id_t: 32 bits
00xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: User key ID range (currently defined)
01xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: Currently defined vendor range
010xxxxxxxxxxxxxxxxxxxxxxxxxxxxx: proposal for 1-bit reduction in do-whatever-you-want-with-it vendor range
011yyyyyyyyyyyyyyyxxxxxxxxxxxxxx: proposal to assign keys from carved out vendor range,
                                  where y = driver lifetime value, and x is key identifier
                                  inside that driver. Length of y and x can be up for debate,
                                  but the field length of the driver location `psa_key_location_t`
                                  inside `psa_key_lifetime_t` should be reduced to match
                                  the length of y in the key ID.
```

**Comment by gilles-peskine-arm** [comment]
The idea with the vendor range was that it was for keys that are “built in” to the core, and a key that's “pre-registered” in a secure element would be effectively registered by a provisioning application and so would be in the application range. The boundary isn't very clear since deployment can vary a lot. If you're both the system vendor and the SE vendor, either mechanism could work.

Mbed TLS doesn't yet support defining “built-in” keys without modifying the library code. I've created https://github.com/ARMmbed/mbedtls/issues/3745 to add support for that.

**Comment by stevew817** [comment]
Resolved through #3774
